Add notification toast (#1295)
sumup-oss/circuit-ui: packages/circuit-ui/components/NotificationToast/NotificationToast.docs.mdx

import { Status, Props, Story } from '../../../../.storybook/components';
import { NotificationToast } from '@sumup/circuit-ui';

# NotificationToast

<Status.Stable />

<Story id="notification-notificationtoast--base" />
<Props />

The notification toast component non-disruptively provides quick and contextual feedback to the user. It can be based on the outcome of an action (i.e. saving a form) or it can pop up without having been triggered explicitly (i.e. a toast message "You made a new sale" showing real time sales in an app).

## When to use it

> The notification toast should be used for a feedback on micro actions (i.e. successfully/unsuccessfully saving some information).

Reviewer: I'm not sure that this section is very useful (duplicates the intro), but let's first come up with a consistent docs structure and clean this up later if necessary

Reviewer 2: I removed this part in the end.

## Usage guidelines

- Be concise, specific and short in the message and always provide a body copy. If needed an optional headline can be included above the body copy.
- The notification toast floats above the content, positioned at the bottom of the screen, centralized horizontally within the content area.

> - For multiple stacked toasts, the new toast slides on top of the previous one. The timeout of the stacked toasts follows chronological order.

Reviewer: These two are not really usage guidelines, they're how the toasts work—developers don't have a choice here. I'd remove them 🙂

> - The auto-dismiss duration is adjustable (default minimum duration is 3000 ms).

Reviewer: I'd remove this since it's already in the accessibility section, and the fact that the duration is adjustable is obvious when looking at the props

> - Toasts can also be dismissed by clicking on a close button in the top-right corner.

Reviewer: This is also obvious IMO.

Generally in other component docs I've been removing these "usage guidelines" sections altogether, they seem very "random" and subjective (e.g. why are we talking about the optional headline but not the different variants here? It doesn't seem consistent/systematic)

Feel free to leave it in here for now, but I think that it's more valuable to write this kind of things either under dedicated sections in "Accessibility" or in the component variations (e.g. variants)

Again, we can revisit this once we come up with a docs template that we can use consistently

## Variants

<Story id="notification-notificationtoast--variants" />

> ### Info

Reviewer: Could we use a list for this instead of headlines?

```md
## Component variations

### Variants

- Use the 'Info' variant for informational, neutral messages
- Use the 'Confirm' variant the successful messages
- ...
```

Reviewer: Also let's add a variation for the toast with headline:

```md
## Component variations

### Toast variants

...

### Toast with headline

...
```

### Confirm

Use the 'Confirm' variant the successful messages.

### Notify

Use the 'Notify' variant for warning messages, and other information a user should be aware of.

### Alert

Use the 'Alert' variant for error messages.

## Usage in code

First, wrap your application in the `ToastProvider` which keeps track of the open toasts and ensures the accessibility of the toast.

```tsx
// _app.tsx for Next.js or App.js for CRA
import { ToastProvider } from '@sumup/circuit-ui';

export default function App() {
  return <ToastProvider>{/* Your content goes here... */}</ToastProvider>;
}
```

Then, use the `useNotificationToast` hook to open a toast from a component:

```tsx
import { useNotificationToast, Button } from '@sumup/circuit-ui';

export function App({}) {
  const { setToast } = useNotificationToast();

  const handleClick = () => {
    setToast({
      body: 'This is a toast message',
    });
  };

  return <Button onClick={handleClick}>Open toast</Button>;
}
```

## Accessibility

By their nature as small quick messages, or interruptions, toasts map to [live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions).

> Live region must always exist on the page at render, before the toast is generated.

Reviewer: For next time: a line break in markdown doesn't do anything, a paragraphs should be on a single line without line breaks 🙂 You can use "toggle word wrap" in VSCode to make it easier to read them though

> Ideally, the toast messages should not contain any interactive elements.

Reviewer: ```suggestion
Toast messages should not contain any interactive elements.
```

> The screen readers are gonna announce the raw text within toast messages, without structure and any need to move the user's focus.

Reviewer: ```suggestion
Screen readers will announce raw text within the toast as it appears, without structure and without moving focus to the toast element.
```

> Keyboard users won’t be able to move to the interactive elements with a clear path.
> For example, there is a toast message with the text “You successfully updated your data.”, followed by a link or button with the text "Back".
> Screen readers are gonna communicate a message as “You successfully updated your data. Back.”, and the semantics of the link "Back" is not announced.

Reviewer: ```suggestion
For example, if a toast read “You successfully updated your data” followed by an "Undo" button, screen readers would announce “You successfully updated your data. Undo”, not allowing keyboard users to interact with the button.
```

To match the content, the role and aria-live level is adapted. The role is set to `"status"`, and it has `aria-live="polite"` meaning the screen reader will wait before announcing the update.

### Best practices

This component adheres to the following principles from the Web Content Accessibility Guidelines (WCAG):

#### Toast duration

The auto-dimiss duration of the toast is adjustable.
Based on how fast the average American reads, a good lenght of time to keep the toast message up is 5 seconds plus 1 extra second for every 120 words.

> The shortest default that should be used as a best practice is 6 seconds.

Reviewer: Can we add a reference here? And if this is true, shouldn't we make our own default more than 3s?

Reviewer 2: I was literally thinking about this too, I think we should then set the default to more then 3s. I will add it in the list to discuss with GC after the holidays.

#### Consistent identification

> Based on the WCAG guideline [3.2.4](https://www.w3.org/WAI/WCAG21/Understanding/consistent-identification.html) each component should be consistent with another.

Reviewer: I don't understand what this means in this context 🙈

#### Location

> Following consistent identificaton, toast messages should be also placed consistently. Generally toast messages usually appear at the start or end of the DOM. If you place them somewhere else, it can cause difficulties for screen reader users to locate them.

Reviewer: This seems wrong—screen reader users probably wouldn't have time to reach the toast in the DOM before it disappears anyways, that's why we're using a live region. I guess that a consistent location at the bottom center of the screen is useful for sighted users, though

#### Keyboard access and focus order

For any interactive element, user should be able to tab and shift through them by using the keyboard, which means they require a keyboard focus indicator.
Follow more in the WCAG guideline [2.1.1](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html)

> Using the keyboard, if the toast message has any interactive elements, then it should always show up in the same place in the focus order, in a way that makes sense to a user.

Reviewer: But the toast should never have interactive elements 🤔

Reviewer 2: Most of the articles cover up the cases when there is interactive element for the keyboard navigation. I agree that it does not go well with the idea of the toasts not having any interactive element. I wouldn't mind deleting this part in the end if that is ok for you ?

Reviewer: Yes I think that makes sense. A toast can be badly accessible if it has interactivity, but it can not be accessible if it has interactivity+auto-dismissal. In our case it has auto-dismissal, therefore it should never have interactivity 🙂 (the close button is just progressive enhancement and doesn't count as interactivity)

import { Status, Props, Story } from '../../../../.storybook/components';
import { NotificationToast } from '@sumup/circuit-ui';

# NotificationToast

<Status.Stable />

<Story id="notification-notificationtoast--base" />
<Props />

> The notification toast component non-disruptively provides quick, contextual or feedback based on the outcome of an action.

Reviewer: This is not quite right: the toast can be triggered by an action (ex. saving a form), but it can also pop up without having been explicitly triggered (e.g. if we showed realtime sales in an app there could be a toast saying "You made a new sale").

This is actually (one of) the reason why toasts are so problematic—if they were only triggered by user interactions, then we could make them non-modal dialogs, send focus to them (this would allow actions on toasts), and send focus back to the element that triggered them on close.

(we may need to update this elsewhere in the docs)

## Variants

<Story id="notification-notificationtoast--variants" />

### Info

Use 'Info' variant for the informational, neutral messages.

### Confirm

Use 'Confirm' variant for the successful messages.

### Notify

Use 'Notify' variant for a warning messages, information to be aware of.

### Alert

Use 'Alert' variant for error messages.

## Usage in code

First, wrap your application in the `ToastProvider` which keeps track of the open toasts and ensures the accessibility of the toast.

```tsx
// _app.tsx for Next.js or App.js for CRA
import { ToastProvider, BaseStyles } from '@sumup/circuit-ui';

export default function App() {
  return (
      <BaseStyles />
      <ToastProvider>{/* Your content goes here... */}</ToastProvider>
  );
}
```

Then, use the `useNotificationToast` hook to open a toast from a component:

```tsx
import { useNotificationToast, Button, Body } from '@sumup/circuit-ui';

export function App({ name }) {
  const { setToast } = useNotificationToast();

  const handleClick = () => {
    setToast({
      body: 'This is a toast message',
      closeButtonLabel: 'Close modal',
    });
  };

  return <Button onClick={handleClick}>Open toast</Button>;
}
```

> ## Accessibility

Reviewer: One thing I'd add to the a11y section is recommendations on the timeout duration. I've read somewhere that the average American reads 240 chars (think a tweet) in 6 seconds, so we might want to recommend to use 6 seconds for a toast of roughly this length (let's do a bit more research about this)

Now that I'm thinking about this, we could even build it in, i.e. give a development-only warning if the toasts receives a body with e.g. 240 characters but only the default 3 seconds, recommending to increase the timeout duration

By their nature as small quick messages, or interruptions, toasts map to [live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions),
and screen readers are gonna automatically announce the raw text within without any need to move the user's focus.
Live region must always exist on the page at render, before the toast is generated.
To ensure that the toast is always announced as a whole, `aria-atomic` should be set to `"true"` (default value is `"false"`).

### Role

The `aria-role` can be adapted based on the context.
For the most generic use cases that provides an updated status, the role can be set to `"status"`. It has a `aria-live="polite"` meaning the screen reader will wait before announcing the update.
If it is an important message, typically time-sensitive, like an error, use `role="alert"`. It has `aria-live="assertive"` which means it will interrupt whatever the user is doing.

> The third role that can be given to the toast messages is `"log"`. It should be used when the order of the messages matters. It has a `aria-live="polite"`.

Reviewer: I'd actually remove the entire paragraph since `aria-live` is set to `polite` internally (I don't think we should support `assertive` at this time).

Since the screen readers are gonna announce the raw text within toast messages, without structure, keyboard users won’t be able to move to the interactive elements with a clear path.
For example, there is a toast message with the text “You successfully updated your data.”, followed by a link or button with the text "Back".
Screen readers are gonna communicate a message as “You successfully updated your data. Back.”, meaning that the semantics of the link "Back" is not announced.
People using a screen reader may think that the "Back" is an interactive element, but they will not know which element to search for.
In this case you will need a modal or dialog with a `role="dialog"`.
Ideally, the toast messages should not contain any interactive elements.

> ### Best practices

Reviewer: Although all of the best practices below do apply to our toast, none of them are actionable by developers. We define the color, the positioning, etc.

So similar to what Connor mentioned above, I'd either remove it altogether (to keep developers focused on what they really have an impact on) or make it very clear that this is Circuit's concern, not theirs

(☝️ and we should do the same in all components.)

Reviewer: I'm tending towards removing it, because many of these guidelines would have to be repeated in basically every component (e.g. color contrast).

This is rather an accessibility statement to me: at some point we'll state somewhere that we are WCAG AA compliant, which implies that we must meet all AA success criteria, including color, focus for interactive elements, etc.

But would like to hear your thoughts about it 🙂

Author: I really like these sections and would like to keep them. This information is very helpful for maintainers, makes our accessibility efforts more visible, and can inspire and educate developers on these principles.

You're right that we should make this context clear. What about one of these statements at the start of the section?

> This component was created with the following accessibility principles in mind:

> This component adheres to the following principles from the 
Web Content Accessibility Guidelines (WCAG):

Reviewer 2: I was just going through these again, and I am more towards keeping it too. I will make it more clear and update it. I like your suggestion Connor for the start.

#### Consistent identification

Based on the WCAG guideline [3.2.4](https://www.w3.org/WAI/WCAG21/Understanding/consistent-identification.html) each component should be consistent with another.

#### Location

Following consistent identificaton, toast messages should be also placed consistently. Generally toast messages usually appear at the start or end of the DOM. If you place them somewhere else, it can cause difficulties for screen reader users to locate them.

#### Keyboard access and focus order

For any interactive element, user should be able to tab and shift through them by using the keyboard, which means they require a keyboard focus indicator.
Follow more in the WCAG guideline [2.1.1](https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html)

Using the keyboard, if the toast message has any interactive elements, then it should always show up in the same place in the focus order, in a way that makes sense to a user.
This is related to the WCAG guideline [2.4.3](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html).

#### Timing adjustable

This is the one that is most relevant to the toast messages.

> Based on the WCAG guideline [2.2.1](https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html), user should be able to choose how long they want the toast messages to be displayed.

Reviewer: This is also probably not relevant for developers using the toast, so I wouldn't add it here.

But I'm adding this extra comment because we still need to take a stand about this SC. There's [a discussion about this](https://github.com/w3c/wcag/issues/1814#issuecomment-988311675) on the wcag repo with details.

IMO in our case we should be exempt from having to meet it. If WGAC states that status messages must also meet 2.2.1 then we might want to add to our upcoming a11y statement that we're aiming at WCAG AA except 2.2.1, because realistically I don't see timing adjustable happening anytime soon—what we'd want instead is to push against using toasts at all.
